runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunction.java

```diff
@@ -176,7 +184,23 @@ private StateRequestHandler getStateRequestHandler(
     } catch (IOException e) {
       throw new RuntimeException("Failed to setup state handler", e);
     }
+
+    // Need to discard the old key's state
+    if (bagUserStateHandlerFactory != null) {
+      bagUserStateHandlerFactory.resetForNewKey();
```

Comment:
Here you are resetting, but further down you create a new instance of `bagUserStateHandlerFactory`.